pallets/pallet-bonded-coins/src/curves.rs

```diff
@@ -0,0 +1,253 @@
+	SquareRootBondingFunction(SquareRootFunctionParameters<F>),
+	LMSR(LMSRFunctionParameters<F>),
+}
+
+pub enum Operation<F> {
+	Mint(Vec<F>),
+	Burn(Vec<F>),
+}
+
+impl<F> Operation<F> {
+	pub fn inner_value(&self) -> &[F] {
+		match self {
+			Operation::Mint(x) => x,
+			Operation::Burn(x) => x,
+		}
+	}
+}
+
+impl<F> Curve<F>
+where
+	F: FixedSigned + PartialOrd<I9F23> + From<I9F23> + ToFixed,
+	<F as Fixed>::Bits: Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign,
+{
+	fn calculate_accumulated_passive_issuance(passive_issuance: &[F]) -> F {
+		passive_issuance
+			.iter()
+			.fold(F::from_num(0), |sum, x| sum.saturating_add(*x))
+	}
+
+	fn calculate_low_high(op: Operation<F>, active_issuance_pre: F, active_issuance_post: F) -> (F, F) {
+		match op {
+			Operation::Burn(passive) => {
+				let accumulated_passive_issuance = Self::calculate_accumulated_passive_issuance(&passive);
+				(
+					active_issuance_post.saturating_add(accumulated_passive_issuance),
+					active_issuance_pre.saturating_add(accumulated_passive_issuance),
+				)
+			}
+			Operation::Mint(passive) => {
+				let accumulated_passive_issuance = Self::calculate_accumulated_passive_issuance(&passive);
+				(
+					active_issuance_pre.saturating_add(accumulated_passive_issuance),
+					active_issuance_post.saturating_add(accumulated_passive_issuance),
+				)
+			}
+		}
+	}
+	pub fn calculate_cost(
+		&self,
+		active_issuance_pre: F,
+		active_issuance_post: F,
+		op: Operation<F>,
+	) -> Result<F, ArithmeticError> {
+		match self {
+			Curve::PolynomialFunction(params) => {
+				let (low, high) = Self::calculate_low_high(op, active_issuance_pre, active_issuance_post);
+
+				params.calculate_costs(low, high)
+			}
+			Curve::SquareRootBondingFunction(params) => {
+				let (low, high) = Self::calculate_low_high(op, active_issuance_pre, active_issuance_post);
+				params.calculate_costs(low, high)
+			}
+			Curve::LMSR(params) => {
+				let passive_issuance_over_e = op
+					.inner_value()
+					.iter()
+					.map(|x| params.calculate_passive_issuance(*x))
+					.collect::<Result<Vec<F>, ArithmeticError>>()?;
+
+				let passive_issuance = passive_issuance_over_e.iter().try_fold(F::from_num(0), |acc, x| {
+					acc.checked_add(*x).ok_or(ArithmeticError::Overflow)
+				})?;
+
+				let lmsr_calc = LMSRCalculation {
+					m: params.m,
+					passive_issuance,
+				};
+				lmsr_calc.calculate_costs(active_issuance_pre, active_issuance_post)
+			}
+		}
+	}
+}
+
+pub trait BondingFunction<F: FixedSigned + PartialOrd> {
+	fn calculate_costs(&self, low: F, high: F) -> Result<F, ArithmeticError>;
+
+	fn square(x: F) -> Result<F, ArithmeticError> {
+		x.checked_mul(x).ok_or(ArithmeticError::Overflow)
+	}
+}
+
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct PolynomialFunctionParameters<F> {
+	pub m: F,
+	pub n: F,
+	pub o: F,
+}
+
+impl<F> BondingFunction<F> for PolynomialFunctionParameters<F>
```

Comment:
Answer based on: [Shouldn't this be implemented for the bonding curve itself, and not its parameters?](https://github.com/KILTprotocol/kilt-node/pull/757/files/4a4174cbc0e795c208551efdfd7f32d1c8cfcbdc#r1808204051). 

I think implementing it for the Curve enum directly is a bit tricky. The LMSR function requires some special treatment. The passive issuance needs some additional processing, before applying it.